components/pill/index.jsx

```diff
@@ -0,0 +1,284 @@
+		 * _Tested with Mocha framework._
+		 * @callback
+		 */
+		onRemove: PropTypes.func,
+
+		/**
+		 * Role to use for the pill. Default role: button.
+		 * _Tested with Mocha framework._
+		 */
+		role: PropTypes.string,
+
+		/**
+		 * This is a way to specify custom contents for the pill in the case a simple text label is not enough.
+		 * _Tested with Mocha framework._
+		 */
+		children: PropTypes.node
+	},
+
+	getDefaultProps () {
+		return {
+			role: 'button',
+			labels: {},
+			assistiveText: {}
+		};
+	},
+
+	render () {
+		const role = this.props.role || (typeof this.props.onClick === 'function' ? 'button' : null);
+		return (
+			/* eslint-disable jsx-a11y/no-static-element-interactions */
+			<span
+				{...this.restProps()}
+				role={role}
+				className={classNames(
+					'slds-pill',
+					{
+						'slds-pill_link': this.renderAsLink(),
+						'slds-has-error': this.props.hasError,
+						'slds-pill_bare': this.props.bare
+					},
+					this.props.className
+				)}
+				onClick={this.props.onClick}
+				onKeyUp={this.props.onRemove ? this.handleKeyUp : null}
+				ref={this.handleRef}
+			>
+				{this.renderIcon()}
+				{this.renderLabel()}
+				{this.renderRemoveIcon()}
+			</span>
+		);
+	},
+
+	renderIcon () {
+		const icon = this.props.icon || this.props.avatar;
+		if (icon) {
+			return <span className="slds-pill__icon_container">{icon}</span>;
+		}
+		return null;
+	},
+
+	renderLabel () {
+		if (this.props.labels.label) {
+			if (this.renderAsLink() || this.props.href) {
+				return (
+					<a
+						href={this.getHref()}
+						className="slds-pill__action"
+						title={this.props.labels.title || this.props.labels.label}
+					>
+						<span className="slds-pill__label">
+							{this.props.labels.label}
+						</span>
+					</a>
+				);
+			}
+			return (
+				<span className="slds-pill__label" title={this.props.labels.title || this.props.labels.label}>
+					{this.props.labels.label}
+				</span>
+			);
+		}
+		return this.props.children;
+	},
+
+	renderRemoveIcon () {
+		if (typeof this.props.onRemove === 'function') {
+			return (
+				<span
+					className="slds-icon_container slds-pill__remove"
+					title={this.props.labels.removeTitle}
+					role="button"
+					onClick={this.props.onRemove}
+					onKeyUp={this.handleKeyUp}
+				>
+					<UtilityIcon
+						style={{ cursor: 'pointer' }}	// remove when fixed by SLDS CSS
+						category="utility"
+						className="slds-icon slds-icon--x-small slds-icon-text-default"
+						name="close"
+					/>
+					<span className="slds-assistive-text">{this.props.assistiveText.remove || this.props.labels.removeTitle}</span>
+				</span>
+			);
+		}
+
+		return null;
+	},
+
+	/**
+	 * Give focus to the Pill component.
+	 */
+	focus () {
+		this.root.focus();
+	},
+
+	/**
+	 * Removes focus from the component.
+	 */
+	blur () {
+		this.root.blur();
+	},
+
+	handleKeyUp (event, ...rest) {
```

Comment:
Most of DSR component events are on keyDown. Historically, in this library I think it has to do with key modifiers (shift, mostly). Do you need keyUp for your use case. Also, `onKeyUp` is missing from the `PropTypes`.

I would prefer a more declarative way of preventing events than with a false return--something like `onRequestRemove` that prevents removal in the callback. If you look at the Combobox `Pill`, you will see:
```
onClickPill: PropTypes.func.isRequired,
onRequestFocus: PropTypes.func.isRequired,
onRequestFocusOnNextPill: PropTypes.func.isRequired,
onRequestFocusOnPreviousPill: PropTypes.func.isRequired,
onRequestRemove: PropTypes.func.isRequired
```
https://github.com/salesforce/design-system-react/blob/master/components/combobox/private/selected-listbox.jsx#L41
also see
https://github.com/salesforce/design-system-react/blob/master/docs/codebase-overview.md#controlled-and-uncontrolled-components

Reply:
1) If all other components rely on keyDown, then I'll change to keyDown. No matter whether that's bad or ok.
2) Will add a prototype
3) I did see this, but it looks unnecessarily complex (horrible: that's javascript, not C++ :-)) to me and very much specific to the context in which the private pill is being used. We can stick to `defaultPrevented` only, for example - wouldn't this be enough?
4) Can you point me to any other components with similar needs other then the private pill?

Reply:
https://github.com/salesforce/design-system-react/search?utf8=%E2%9C%93&q=onRequestClose&type=

When I was writing Datepicker a year ago, I realized that I needed to know if the event occurred before or after the event mentioned `focus`, `close`, `open`, etc. It's using the name of the event to signal what you are to do in it. Otherwise, I was stuck looking into the source code to know if `onClose` was what I was supposed to do once the menu closed (like focus back on the input) or whether `onClose` was supposed to actually close the menu in the state container. If order doesn't matter, then you can do both in the callback. I think separating these are going to be a good best practice as rendering becomes more asynchronous.